modules/scheduler/scheduler/src/main/java/org/motechproject/scheduler/service/impl/MotechSchedulerServiceImpl.java

```diff
@@ -705,12 +576,256 @@ protected void logObjectIfNotNull(Object obj) {
+        logObjectIfNotNull(schedulableJob);
+
+        assertArgumentNotNull("RunOnceSchedulableJob", schedulableJob);
+        MotechEvent motechEvent = schedulableJob.getMotechEvent();
+
+        DateTime jobStartDate = schedulableJob.getStartDate();
+        assertArgumentNotNull("Job start date", jobStartDate);
+        DateTime currentDate = DateUtil.now();
+        if (jobStartDate.isBefore(currentDate)) {
+            String errorMessage = "Invalid RunOnceSchedulableJob. The job start date can not be in the past. \n" +
+                    " Job start date: " + jobStartDate.toString() +
+                    " Attempted to schedule at:" + currentDate.toString();
+            throw new IllegalArgumentException(errorMessage);
+        }
+
+        JobId jobId = new RunOnceJobId(motechEvent);
+        JobDetail jobDetail = newJob(MotechScheduledJob.class)
+                .withIdentity(jobId.value(), JOB_GROUP_NAME)
+                .build();
+
+        putMotechEventDataToJobDataMap(jobDetail.getJobDataMap(), motechEvent);
+        jobDetail.getJobDataMap().put(UI_DEFINED, schedulableJob.isUiDefined());
+
+        SimpleScheduleBuilder simpleSchedule = simpleSchedule()
+                .withRepeatCount(0)
+                .withIntervalInSeconds(0)
+                .withMisfireHandlingInstructionFireNow();
+
+        Trigger trigger = newTrigger()
+                .withIdentity(triggerKey(jobId.value(), JOB_GROUP_NAME))
+                .forJob(jobDetail)
+                .withSchedule(simpleSchedule)
+                .startAt(DateUtil.toDate(jobStartDate))
+                .build();
+
+        scheduleJob(jobDetail, trigger, update);
+    }
+
+    private void scheduleDayOfWeekJob(DayOfWeekSchedulableJob dayOfWeekSchedulableJob, boolean update) {
+        logObjectIfNotNull(dayOfWeekSchedulableJob);
+
+        MotechEvent motechEvent = dayOfWeekSchedulableJob.getMotechEvent();
+        LocalDate start = dayOfWeekSchedulableJob.getStartDate();
+        LocalDate end = dayOfWeekSchedulableJob.getEndDate();
+        Time time = dayOfWeekSchedulableJob.getTime();
+
+        CronScheduleBuilder cronScheduleBuilder = CronScheduleBuilder.atHourAndMinuteOnGivenDaysOfWeek(time.getHour(),
+                time.getMinute(), dayOfWeekSchedulableJob.getCronDays()
+                        .toArray(new Integer[dayOfWeekSchedulableJob.getCronDays().size()]));
+
+        CronTriggerImpl cronTrigger = (CronTriggerImpl) cronScheduleBuilder.build();
+        CronSchedulableJob cronSchedulableJob = new CronSchedulableJob(motechEvent, cronTrigger.getCronExpression(),
+                start.toDateMidnight().toDateTime(), end != null ? end.toDateMidnight().toDateTime() : null,
+                dayOfWeekSchedulableJob.isIgnorePastFiresAtStart(), dayOfWeekSchedulableJob.isUiDefined());
+
+        scheduleCronJob(cronSchedulableJob, update);
+    }
+
+    private void scheduleJob(JobDto dto, boolean update) {
```

Comment:
we shouldn't clutter this service too much - translating and validating from the dto to the scheduler service calls should be done by a different component.